back/boxtribute_server/business_logic/statistics/crud.py

```diff
@@ -206,3 +215,53 @@ def compute_top_products_donated(base_id):
 
     dimensions = _generate_dimensions("category", "product", "size", facts=facts)
     return {"facts": facts, "dimensions": dimensions}
+
+
+TARGET_ID_SEPARATOR = "---"
+
+
+def compute_moved_boxes(base_id):
+    """Count all boxes moved to locations in the given base, grouped by date of
+    movement, location, base, product category, and box state.
+    """
+    # TODO: use more precise query with box versions
+    # This selects only information of boxes moved from InStock to Donated state
+    selection = (
+        DbChangeHistory.select(
+            fn.MAX(DbChangeHistory.change_date).alias("moved_on"),
+            fn.CONCAT(Base.name, TARGET_ID_SEPARATOR, Location.name).alias("target_id"),
+            Product.category.alias("category_id"),
+            fn.COUNT(Box.id).alias("boxes_count"),
+        )
+        .join(
+            Box,
+            on=(
+                (DbChangeHistory.record_id == Box.id)
+                & (DbChangeHistory.table_name == Box._meta.table_name)
+                & (DbChangeHistory.changes == Box.state.column_name)
+                & (DbChangeHistory.from_int == BoxState.InStock)
+                & (DbChangeHistory.to_int == BoxState.Donated)
+            ),
+        )
+        .join(
+            Product,
+            on=((Box.product == Product.id) & (Product.base == base_id)),
+        )
+        .join(
+            Location,
+            src=Box,
+            on=((Box.location == Location.id) & (Location.base == base_id)),
+        )
+        .join(Base)
```

Comment:
don't you need an `on` clause here?

Reply:
peewee is able to auto-derive this because the previous model (Location) has a single FK reference to the Base model.